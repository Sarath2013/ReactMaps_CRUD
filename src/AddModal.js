import React, { Component } from 'react';
import LocationDetials from './LocationDetials'
import axios from 'axios';
import PropTypes from 'prop-types';

class AddModal extends Component {
    constructor(props) {
        super(props);
        this.state = {
            location: props.curLocation,
            searchValue: props.curLocation.name,
            disableBtn: true,
            noGeoData: false,
            disableGeoBtn: true,
            isDuplicate: false
        };
        this.addMapLocation = this.addMapLocation.bind(this);
        this.editMapLocation = this.editMapLocation.bind(this);
        this.searchChange = this.searchChange.bind(this);
        this.getGeoCordiantes = this.getGeoCordiantes.bind(this);
    }

    searchChange(event) {
        let disableGeoBtn = true;
        if ((this.props.modalName === "Add" && event.target.value) || (this.props.modalName === "Edit" && event.target.value && event.target.value !== this.state.searchValue))
            disableGeoBtn = false;
        this.setState({ searchValue: event.target.value, hasError: false, disableGeoBtn });
    }

    addMapLocation() {
        axios.post('http://localhost:4000/createMarker', this.state.location, { headers: { "Access-Control-Allow-Origin": "*" } })
            .then(res => {
                this.setState({ hasError: false });
                this.props.toggle(res.data.location);
            })
            .catch(error => {
                console.log(error);
                this.setState({ hasError: true });
            })
    }

    editMapLocation() {
        axios.put(`http://localhost:4000/editMarker/${this.props.curLocation._id}`, this.state.location, { headers: { "Access-Control-Allow-Origin": "*" } })
            .then(res => {
                let location = { ...this.props.curLocation };
                location.name = this.state.location.name;
                location.lat = this.state.location.lat;
                location.lng = this.state.location.lng;
                this.setState({ hasError: false });
                this.props.toggle(location);
            })
            .catch(error => {
                console.log(error);
                this.setState({ hasError: true });
            })
    }

    getGeoCordiantes() {
        axios.get(`https://maps.googleapis.com/maps/api/geocode/json?address=${this.state.searchValue}&key=${process.env.REACT_APP_MAPS_API_KEY}`)
            .then(res => {
                const data = res.data;
                if (data.results.length > 0) {
                    let filterArr = this.props.locations.filter((obj) => obj.name === data.results[0].formatted_address);
                    if (data.results[0].formatted_address === this.props.curLocation.name) {
                        this.setState({ disableGeoBtn: true, noGeoData: false });
                    } else if (filterArr.length > 0) {
                        this.setState({ isDuplicate: true, noGeoData: false });
                    } else {
                        this.setState({
                            location: {
                                name: data.results[0].formatted_address,
                                lat: data.results[0].geometry.location.lat,
                                lng: data.results[0].geometry.location.lng,
                            },
                            noGeoData: false,
                            disableBtn: false,
                            isDuplicate: false
                        });
                    }
                } else {
                    this.setState({
                        noGeoData: true,
                        isDuplicate: false
                    });
                }
            })
            .catch(error => {
                console.log(error)
                this.setState({ hasError: true, disableBtn: true, isDuplicate: false });
            })
    }

    render() {
        return (
            <div
                className="modal show"
                tabIndex="-1"
                role="dialog"
                aria-hidden="true"
                style={{ display: "block" }}
            >
                <div className="modal-dialog" role="document">
                    <div className="modal-content">
                        <div className="modal-header">
                            <h3>{this.props.modalName} Marker</h3>
                            <button
                                type="button"
                                className="close"
                                aria-label="Close"
                                onClick={this.props.toggleClose}
                            >
                                <span aria-hidden="true">&times;</span>
                            </button>
                        </div>
                        <div className="modal-body">
                            <form>
                                <div className="form-group">
                                    <input name="searchInput" type="text" className="form-control" placeholder="Type Place" value={this.state.searchValue} onChange={this.searchChange} />
                                </div>
                                <button type="button" className="btn btn-primary geo-btn" disabled={this.state.disableGeoBtn} onClick={this.getGeoCordiantes}>
                                    Get Geo Cordinates</button>
                            </form>
                            {(this.state.location.name) ?
                                <LocationDetials location={this.state.location} /> : null}
                            {(this.state.noGeoData) ? (<div className="error-msg locations">
                                No locations found.
                                    </div>) : null
                            }
                            {(this.state.hasError) ?
                                <div className="error-msg">Something went wrong. Please try agian later.</div> : null}
                            {(this.state.isDuplicate) ?
                                <div className="error-msg exists">Marker already exists there.</div> : null}
                        </div>
                        <div className="modal-footer">
                            <button type="button" className="btn btn-secondary" onClick={this.props.toggleClose}>
                                Close</button>
                            <button type="button" className="btn btn-primary" disabled={this.state.disableBtn} onClick={(this.props.modalName === "Add") ? this.addMapLocation : this.editMapLocation}>
                                {this.props.modalName}</button>
                        </div>
                    </div>
                </div>
            </div>
        );
    }
}


AddModal.propTypes = {
    toggle: PropTypes.func.isRequired,
    toggleClose: PropTypes.func.isRequired,
    modalName: PropTypes.string.isRequired,
    curLocation: PropTypes.shape({
        name: PropTypes.string,
        lat: PropTypes.number,
        lng: PropTypes.number
    }).isRequired,
    locations: PropTypes.arrayOf(PropTypes.shape({
        name: PropTypes.string.isRequired,
        lat: PropTypes.number.isRequired,
        lng: PropTypes.number.isRequired
    })).isRequired,
}

export default AddModal;