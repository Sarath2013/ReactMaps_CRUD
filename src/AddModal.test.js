import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import AddModal from './AddModal';

const addProps = {
    toggle: jest.fn(),
    toggleClose: jest.fn(),
    modalName: "Add",
    curLocation: {
    },
    locations: []
};

const editProps = {
    toggle: jest.fn(),
    toggleClose: jest.fn(),
    modalName: "Edit",
    curLocation: {
        name: 'Berlin',
        lat: 52.52,
        lng: 13.405
    },
    locations: [{
        name: 'Berlin',
        lat: 52.52,
        lng: 13.405
    }]
};
let wrapper;

describe("Rendering AddModal", () => {
    beforeAll(() => {
    });

    afterAll(() => {
        wrapper = null;
    });

    test('renders snapshot', () => {
        wrapper = shallow(<AddModal {...addProps} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    test('Check AddModal Elements', () => {
        wrapper = mount(<AddModal {...addProps} />);
        wrapper.update();
        expect(wrapper.find('.modal-header h3').text()).toBe("Add Marker");
        expect(wrapper.find('.geo-btn').text()).toBe("Get Geo Cordinates");
        expect(wrapper.find('.loc-title').length).toBe(0);
        expect(wrapper.find('.loc-name').length).toBe(0);
        expect(wrapper.find('.loc-lat').length).toBe(0);
        expect(wrapper.find('.loc-lng').length).toBe(0);
        expect(wrapper.find('.modal-footer .btn-secondary').text()).toBe("Close");
        expect(wrapper.find('.modal-footer .btn-primary').text()).toBe("Add");
        wrapper.unmount();
    });
});

describe("Rendering EditModal", () => {
    beforeAll(() => {
    });

    afterAll(() => {
        wrapper = null;
    });

    test('renders snapshot', () => {
        wrapper = shallow(<AddModal {...editProps} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    test('Check AddModal Elements', () => {
        wrapper = mount(<AddModal {...editProps} />);
        wrapper.update();
        expect(wrapper.find('.modal-header h3').text()).toBe("Edit Marker");
        expect(wrapper.find('input.form-control').props().value).toBe("Berlin");
        expect(wrapper.find('.geo-btn').text()).toBe("Get Geo Cordinates");
        expect(wrapper.find('.loc-title').text()).toBe('Berlin');
        expect(wrapper.find('.loc-name').text()).toBe('Berlin');
        expect(wrapper.find('.loc-lat').text()).toBe('latitude: 52.52');
        expect(wrapper.find('.loc-lng').text()).toBe('longitude: 13.405');
        expect(wrapper.find('.modal-footer .btn-secondary').text()).toBe("Close");
        expect(wrapper.find('.modal-footer .btn-primary').text()).toBe("Edit");
        wrapper.unmount();
    });
});