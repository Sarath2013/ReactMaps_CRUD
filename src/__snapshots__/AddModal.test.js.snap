// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering AddModal renders snapshot 1`] = `
<div
  aria-hidden="true"
  className="modal show"
  role="dialog"
  style={
    Object {
      "display": "block",
    }
  }
  tabIndex="-1"
>
  <div
    className="modal-dialog"
    role="document"
  >
    <div
      className="modal-content"
    >
      <div
        className="modal-header"
      >
        <h3>
          Add
           Marker
        </h3>
        <button
          aria-label="Close"
          className="close"
          onClick={[MockFunction]}
          type="button"
        >
          <span
            aria-hidden="true"
          >
            ×
          </span>
        </button>
      </div>
      <div
        className="modal-body"
      >
        <form>
          <div
            className="form-group"
          >
            <input
              className="form-control"
              name="searchInput"
              onChange={[Function]}
              placeholder="Type Place"
              type="text"
            />
          </div>
          <button
            className="btn btn-primary geo-btn"
            disabled={true}
            onClick={[Function]}
            type="button"
          >
            Get Geo Cordinates
          </button>
        </form>
      </div>
      <div
        className="modal-footer"
      >
        <button
          className="btn btn-secondary"
          onClick={[MockFunction]}
          type="button"
        >
          Close
        </button>
        <button
          className="btn btn-primary"
          disabled={true}
          onClick={[Function]}
          type="button"
        >
          Add
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Rendering EditModal renders snapshot 1`] = `
<div
  aria-hidden="true"
  className="modal show"
  role="dialog"
  style={
    Object {
      "display": "block",
    }
  }
  tabIndex="-1"
>
  <div
    className="modal-dialog"
    role="document"
  >
    <div
      className="modal-content"
    >
      <div
        className="modal-header"
      >
        <h3>
          Edit
           Marker
        </h3>
        <button
          aria-label="Close"
          className="close"
          onClick={[MockFunction]}
          type="button"
        >
          <span
            aria-hidden="true"
          >
            ×
          </span>
        </button>
      </div>
      <div
        className="modal-body"
      >
        <form>
          <div
            className="form-group"
          >
            <input
              className="form-control"
              name="searchInput"
              onChange={[Function]}
              placeholder="Type Place"
              type="text"
              value="Berlin"
            />
          </div>
          <button
            className="btn btn-primary geo-btn"
            disabled={true}
            onClick={[Function]}
            type="button"
          >
            Get Geo Cordinates
          </button>
        </form>
        <LocationDetials
          location={
            Object {
              "lat": 52.52,
              "lng": 13.405,
              "name": "Berlin",
            }
          }
        />
      </div>
      <div
        className="modal-footer"
      >
        <button
          className="btn btn-secondary"
          onClick={[MockFunction]}
          type="button"
        >
          Close
        </button>
        <button
          className="btn btn-primary"
          disabled={true}
          onClick={[Function]}
          type="button"
        >
          Edit
        </button>
      </div>
    </div>
  </div>
</div>
`;
